libraries/botbuilder-dialogs/src/prompts/choicePrompt.ts

```diff
@@ -119,10 +119,9 @@ export class ChoicePrompt extends Prompt<FoundChoice> {
     }
 
     private determineCulture(activity: Activity, opt: FindChoicesOptions = null): string {
-        const optLocale = opt && opt.locale ? opt.locale : null;
-        let culture = PromptCultureModels.mapToNearestLanguage(activity.locale || optLocale || this.defaultLocale || PromptCultureModels.English.locale);
-        if (!culture || !this.choiceDefaults[culture])
-        {
+        const optLocale: string = opt && opt.locale ? opt.locale : null;
+        let culture: string = PromptCultureModels.mapToNearestLanguage(activity.locale || optLocale || this.defaultLocale);
+        if (!culture || !this.choiceDefaults.hasOwnProperty(culture)) {
```

Comment:
Your changes make sense (thank you for removing the default `PromptCultureModels.English.locale` value).

However, this function in the larger context doesn't make sense to me. Could you walk me through it?

In [ChoicePrompt.onRecognize](https://github.com/microsoft/botbuilder-js/blob/499790d51a0d1567c06a9b38e883ef5134b68c48/libraries/botbuilder-dialogs/src/prompts/choicePrompt.ts#L111) determineCulture is called and overrides any pre-existing values on `this.recognizerOptions`. Aren't there concurrency issues if either `this.choiceDefaults` or `this.recognizerOptions` changes?

(`this.choiceDefaults` is potentially a pointer to a public object and `this.recognizerOptions` is a public property)

In [ChoicePrompt.onPrompt](https://github.com/microsoft/botbuilder-js/blob/499790d51a0d1567c06a9b38e883ef5134b68c48/libraries/botbuilder-dialogs/src/prompts/choicePrompt.ts#L86) the locale is obtained without it always being used. 

If `this.choiceOptions || this.choiceDefaults[PromptCultureModels.English.locale]` resolve to undefined, this is fine because [ChoiceFactory provides default values instead](https://github.com/microsoft/botbuilder-js/blob/499790d51a0d1567c06a9b38e883ef5134b68c48/libraries/botbuilder-dialogs/src/choices/choiceFactory.ts#L167-L173)?

It seems like there are some redundancies going on, so any insight (and test coverage) would be helpful at clearing things up

Reply:
I think `determineCulture` is just supposed to determine the culture by accounting for potential missing or malformed values and resorting to fallbacks.

`this.choiceDefaults` is private and doesn't change. Sure the object is public like you said, but it would be pretty strange to modify the object passed as choice defaults, and I don't think anything bad would happen even if it does get modified. Like with `this.recognizerOptions`, the worst that could happen is that different settings get used. It's certainly strange that the options locale gets modified in `onRecognize` but if the property was undefined then the new object that's created doesn't get assigned to the property. I guess if the same dialog instance was used by multiple users with different locales then their locales could get passed from one to another, and I guess that's why the options locale is third in the fallback chain.

I can see that in `onPrompt` the locale is obtained without it always being used, and I suspect the author chose clarity over efficiency since `determineCulture` isn't a slow function to run.

Yes, it looks like `ChoiceFactory` provides default values, but that seems a little outside the scope of this PR. The scope has already creeped beyond the customer's issue so I had thought it was more likely that you'd think it was too much than too little. While I was happy to expand the scope, I had still thought this PR should just be about determining the culture. Do you want to also include code about making sure the choice options work properly?